Compilation test failures with Scala 2 CC library TASTy

#19603 fixes the basics and we can now compile most tests. 

I tried to go further with
```
sbt> set ThisBuild/Build.scala2Library := Build.Scala2LibraryCCTasty
sbt> scala3-bootstrapped/testCompilation
```
Here's the current list of failures:
```
32 suites passed, 4 failed, 36 total
    tests/neg/i8988.scala failed
    tests/neg/abstract-givens.scala failed
    tests/neg/i18682.scala failed
    tests/run/enrich-gentraversable.scala failed
    tests/run/t5328.scala failed
    tests/run/xml.scala failed
    tests/run/t4332b.scala failed
    tests/run/abstract-givens.scala failed
    tests/run/t2552.scala failed
    tests/run/t2544.scala failed
    tests/run/i16459.scala failed
    tests/run/t6114.scala failed
    tests/pos/i18345.scala failed
    tests/run/t3269.scala failed
    tests/pos/spec-asseenfrom.scala failed
    tests/pos/i19009.case1.scala failed
    tests/pos/i14966a.scala failed
    tests/run/iterables.scala failed
    tests/run/i14582.scala failed
    tests/run/t4054.scala failed
    tests/pos/i14953.scala failed
    tests/run/t5201.scala failed
    tests/run/t6968.scala failed
    tests/run/range-unit.scala failed
    tests/pos/t5120.scala failed
    tests/run/t6271.scala failed
    tests/run/enrich-gentraversable.scala failed
    tests/pos/i1036.scala failed
    tests/pos/t6925.scala failed
    tests/run/Course-2002-10.scala failed
    tests/pos/tyql.scala failed
    tests/run/t4288.scala failed
    tests/run/main-annotation-newMain.scala failed
    tests/run/Pouring.scala failed
    tests/run/t5328.scala failed
    tests/run/t6628.scala failed
    tests/run/arrayview.scala failed
    tests/run/colltest1.scala failed
    tests/pos/gadt-accumulatable.scala failed
    tests/run/t3397.scala failed
    tests/run/t2544.scala failed
    tests/run/i16459.scala failed
    tests/run/t3516.scala failed
    tests/pos/tcpoly_ticket2096.scala failed
    tests/run/t4582.scala failed
    tests/run/i4523.scala failed
    tests/run/iterators.scala failed
    tests/run/stream-stack-overflow-filter-map.scala failed
    tests/pos/t4716.scala failed
    tests/run/streamWithFilter.scala failed
    tests/run/IArrayOps.scala failed
    tests/run/ReverseSeqView.scala failed
    tests/run/viewtest.scala failed
    tests/run/numbereq.scala failed
    tests/run/t3760.scala failed
    tests/run/distinct.scala failed
    tests/run/t4332b.scala failed
    tests/pos/i6199a.scala failed
    tests/run/view-headoption.scala failed
    tests/run/abstract-givens.scala failed
    tests/pos/i9967.scala failed
    tests/pos/i1777.scala failed
    tests/pos/StringContext.scala failed
    tests/run/t4954.scala failed
    tests/pos/i5418.scala failed
    tests/run/t6584.scala failed
    tests/run/extmethods2.scala failed
    tests/pos/t0068.scala failed
    tests/run/kmpSliceSearch.scala failed
    tests/run/patmat-bind-typed.scala failed
    tests/run/t1220.scala failed
    tests/run/t5867.scala failed
    tests/pos/syntaxRewrites.scala failed
    tests/run/forvaleq.scala failed
    tests/pos/t6205.scala failed
    tests/run/Pouring2.scala failed
    tests/run/unittest_iterator.scala failed
    tests/pos/indent-colons.scala failed
    tests/pos/t4717.scala failed
    tests/run/t2552.scala failed
    tests/run/lst failed
    tests/run/t5053.scala failed
    tests/run/lists-run.scala failed
    tests/run/t1074.scala failed
    tests/run/seqlike-kmp.scala failed
    tests/pos/tyql.scala failed
    tests/pos/t6925.scala failed
```
We need to follow up on this next.

_Originally posted by @odersky in https://github.com/lampepfl/dotty/issues/19586#issuecomment-1925328488_

Orphan parameter while Ychecking with CC library.

Minimization of `tests/pos/t6925.scala`

```scala
class Test:
  def f(xs: Set[Int]) = xs collect { case x => x }
  def g(xs: Set[Int]): Set[Int] = f(xs)
```

```
sbt> set ThisBuild/Build.scala2Library := Build.Scala2LibraryCCTasty
sbt> scala3-bootstrapped/scalac Test.scala -Ycheck:all
```

```scala
checking tests/pos/t6925.scala after phase typer

  exception while retyping def g(xs: Set[Int]): Set[Int] = this.f(xs) of class DefDef # -1

  An unhandled exception was thrown in the compiler.
  Please file a crash report here:
  https://github.com/lampepfl/dotty/issues/new/choose
  For non-enriched exceptions, compile with -Yno-enrich-error-messages.

     while compiling: tests/pos/t6925.scala
        during phase: Ycheck
                mode: Mode(ImplicitsEnabled)
     library version: version 2.13.12
    compiler version: version 3.4.1-RC1-bin-SNAPSHOT-git-7bcfce2
            settings: -Vprint List(typer) -Ycheck List(all) -classpath /Users/nicolasstucki/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/scala-lang/scala-library/2.13.12/scala-library-2.13.12.jar:/Users/nicolasstucki/Github/dotty/library/../out/bootstrap/scala3-library-bootstrapped/scala-3.4.1-RC1-bin-SNAPSHOT-nonbootstrapped/scala3-library_3-3.4.1-RC1-bin-SNAPSHOT.jar:/Users/nicolasstucki/Github/dotty/scala2-library-cc-tasty/../out/bootstrap/scala2-library-cc-tasty/scala-3.4.1-RC1-bin-SNAPSHOT-nonbootstrapped/scala2-library-cc-tasty_3-3.4.1-RC1-bin-SNAPSHOT.jar -d /

*** error while checking tests/pos/t6925.scala after phase typer ***
Exception in thread "main" java.lang.AssertionError: assertion failed: orphan param: xs.type, hash of binder = 1664998412, tree = <empty>, type = AnnotatedType(AppliedType(TypeRef(TermRef(ThisType(TypeRef(NoPrefix,module class collection)),object immutable),trait Set),List(TypeRef(TermRef(ThisType(TypeRef(NoPrefix,module class <root>)),object scala),class Int))),ConcreteAnnotation(Apply(Select(New(Select(Select(Select(Ident(_root_),scala),annotation),retains)),<init>),List(Typed(SeqLiteral(List(This(Ident()), Ident(pf)),TypeTree[TypeRef(TermRef(ThisType(TypeRef(NoPrefix,module class <root>)),object scala),Any)]),TypeTree[AppliedType(TypeRef(TermRef(ThisType(TypeRef(NoPrefix,module class <root>)),object scala),<repeated>),List(TypeRef(TermRef(ThisType(TypeRef(NoPrefix,module class <root>)),object scala),Any)))])))))
        at scala.runtime.Scala3RunTime$.assertFailed(Scala3RunTime.scala:8)
        at dotty.tools.dotc.transform.TreeChecker$$anon$2.apply(TreeChecker.scala:172)
        at dotty.tools.dotc.core.Annotations$Annotation$$anon$1.apply(Annotations.scala:64)
        ...
```

--- a/compiler/src/dotty/tools/dotc/printing/PlainPrinter.scala
+++ b/compiler/src/dotty/tools/dotc/printing/PlainPrinter.scala
@@ -242,7 +242,7 @@ class PlainPrinter(_ctx: Context) extends Printer {
         val refsText = if showAsCap then rootSetText else toTextCaptureSet(refs)
         toTextCapturing(parent, refsText, boxText)
       case tp @ RetainingType(parent, refs) =>
-        if ctx.compilationUnit.needsCaptureChecking then
+        if Feature.ccEnabledSomewhere then
           val refsText = refs match
             case ref :: Nil if ref.symbol == defn.captureRoot => rootSetText
             case _ => toTextRetainedElems(refs)

--- a/compiler/src/dotty/tools/dotc/printing/PlainPrinter.scala
+++ b/compiler/src/dotty/tools/dotc/printing/PlainPrinter.scala
@@ -242,10 +242,12 @@ class PlainPrinter(_ctx: Context) extends Printer {
         val refsText = if showAsCap then rootSetText else toTextCaptureSet(refs)
         toTextCapturing(parent, refsText, boxText)
       case tp @ RetainingType(parent, refs) =>
-        val refsText = refs match
-          case ref :: Nil if ref.symbol == defn.captureRoot => rootSetText
-          case _ => toTextRetainedElems(refs)
-        toTextCapturing(parent, refsText, "") ~ Str("R").provided(printDebug)
+        if ctx.compilationUnit.needsCaptureChecking then
+          val refsText = refs match
+            case ref :: Nil if ref.symbol == defn.captureRoot => rootSetText
+            case _ => toTextRetainedElems(refs)
+          toTextCapturing(parent, refsText, "") ~ Str("R").provided(printDebug)
+        else toText(parent)
       case tp: PreviousErrorType if ctx.settings.XprintTypes.value =>
         "<error>" // do not print previously reported error message because they may try to print this error type again recuresevely
       case tp: ErrorType =>

--- a/scala2-library-cc/src/scala/collection/Seq.scala
+++ b/scala2-library-cc/src/scala/collection/Seq.scala
@@ -957,7 +957,7 @@ trait SeqOps[+A, +CC[_], +C] extends Any with SeqViewOps[A, CC, C] { self =>
     *                                    lazy collection this exception may be thrown at a later time or not at
     *                                    all (if the end of the collection is never evaluated).
     */
-  def updated[B >: A](index: Int, elem: B): CC[B] = {
+  override def updated[B >: A](index: Int, elem: B): CC[B] = {
     if(index < 0) throw new IndexOutOfBoundsException(index.toString)
     val k = knownSize
     if(k >= 0 && index >= k) throw new IndexOutOfBoundsException(index.toString)

--- a/scala2-library-cc/src/scala/collection/Seq.scala
+++ b/scala2-library-cc/src/scala/collection/Seq.scala
@@ -206,7 +206,7 @@ trait SeqOps[+A, +CC[_], +C] extends Any with SeqViewOps[A, CC, C] { self =>
     *
     * @return a new $coll consisting of all the elements of this $coll without duplicates.
     */
-  def distinct: C = distinctBy(identity)
+  override def distinct: C = distinctBy(identity)
 
   /** Selects all the elements of this $coll ignoring the duplicates as determined by `==` after applying
     * the transforming function `f`.
@@ -215,7 +215,7 @@ trait SeqOps[+A, +CC[_], +C] extends Any with SeqViewOps[A, CC, C] { self =>
     * @tparam B the type of the elements after being transformed by `f`
     * @return a new $coll consisting of all the elements of this $coll without duplicates.
     */
-  def distinctBy[B](f: A -> B): C = fromSpecific(new View.DistinctBy(this, f))
+  override def distinctBy[B](f: A -> B): C = fromSpecific(new View.DistinctBy(this, f))
 
   /** Returns new $coll with elements in reversed order.
    *
@@ -293,7 +293,7 @@ trait SeqOps[+A, +CC[_], +C] extends Any with SeqViewOps[A, CC, C] { self =>
    *          all elements of this $coll followed by the minimal number of occurrences of `elem` so
    *          that the resulting collection has a length of at least `len`.
    */
-  def padTo[B >: A](len: Int, elem: B): CC[B] = iterableFactory.from(new View.PadTo(this, len, elem))
+  override def padTo[B >: A](len: Int, elem: B): CC[B] = iterableFactory.from(new View.PadTo(this, len, elem))
 
   /** Computes the length of the longest segment that starts from the first element
     *  and whose elements all satisfy some predicate.
@@ -544,7 +544,7 @@ trait SeqOps[+A, +CC[_], +C] extends Any with SeqViewOps[A, CC, C] { self =>
    *    // List(b, b, a)
    *  }}}
    */
-  def permutations: Iterator[C] =
+  override def permutations: Iterator[C] =
     if (isEmpty) Iterator.single(coll)
     else new PermutationsItr
 
@@ -585,7 +585,7 @@ trait SeqOps[+A, +CC[_], +C] extends Any with SeqViewOps[A, CC, C] { self =>
    *    // List(b, a)
    *  }}}
    */
-  def combinations(n: Int): Iterator[C] =
+  override def combinations(n: Int): Iterator[C] =
     if (n < 0 || n > size) Iterator.empty
     else new CombinationsItr(n)
 
@@ -759,7 +759,7 @@ trait SeqOps[+A, +CC[_], +C] extends Any with SeqViewOps[A, CC, C] { self =>
    *    List("Bobby", "Bob", "John", "Steve", "Tom")
    *  }}}
    */
-  def sortWith(lt: (A, A) => Boolean): C = sorted(Ordering.fromLessThan(lt))
+  override def sortWith(lt: (A, A) => Boolean): C = sorted(Ordering.fromLessThan(lt))
 
   /** Sorts this $coll according to the Ordering which results from transforming
     * an implicitly given Ordering with a transformation function.
@@ -786,7 +786,7 @@ trait SeqOps[+A, +CC[_], +C] extends Any with SeqViewOps[A, CC, C] { self =>
     *    res0: Array[String] = Array(The, dog, fox, the, lazy, over, brown, quick, jumped)
     *  }}}
     */
-  def sortBy[B](f: A => B)(implicit ord: Ordering[B]): C = sorted(ord on f)
+  override def sortBy[B](f: A => B)(implicit ord: Ordering[B]): C = sorted(ord on f)
 
   /** Produces the range of all indices of this sequence.
     * $willForceEvaluation
@@ -944,7 +944,7 @@ trait SeqOps[+A, +CC[_], +C] extends Any with SeqViewOps[A, CC, C] { self =>
     *                   except that `replaced` elements starting from `from` are replaced
     *                   by all the elements of `other`.
     */
-  def patch[B >: A](from: Int, other: IterableOnce[B]^, replaced: Int): CC[B] =
+  override def patch[B >: A](from: Int, other: IterableOnce[B]^, replaced: Int): CC[B] =
     iterableFactory.from(new View.Patched(this, from, other, replaced))
       .unsafeAssumePure // assume pure OK since iterableFactory.from is eager for Seq
       

--- a/scala2-library-cc/src/scala/collection/SeqView.scala
+++ b/scala2-library-cc/src/scala/collection/SeqView.scala
@@ -30,10 +30,16 @@ trait SeqViewOps[+A, +CC[_], +C] extends Any with IterableOps[A, CC, C] {
   def length: Int
   def apply(x: Int): A
   def appended[B >: A](elem: B): CC[B]^{this}
-  def updated[B >: A](index: Int, elem: B): CC[B]^{this}
   def prepended[B >: A](elem: B): CC[B]^{this}
   def reverse: C^{this}
   def sorted[B >: A](implicit ord: Ordering[B]): C^{this}
+
+  // Placeholder implementation for that method in SeqOps.
+  // This is needed due to the change in the class hierarchy in cc stdlib.
+  // See #19660 and #19819.
+  def updated[B >: A](index: Int, elem: B): CC[B]^{this} =
+    assert(false, "This is a placeholder implementation in the capture checked Scala 2 library.")
+    ???
 
   def reverseIterator: Iterator[A]^{this} = reversed.iterator
 }
@@ -45,15 +51,6 @@ trait SeqView[+A] extends SeqViewOps[A, View, View[A]] with View[A] {
 
   override def map[B](f: A => B): SeqView[B]^{this, f} = new SeqView.Map(this, f)
   override def appended[B >: A](elem: B): SeqView[B]^{this} = new SeqView.Appended(this, elem)
-
-  // Copied from SeqOps. This is needed due to the change of class hierarchy in stdlib.
-  // See #19660.
-  override def updated[B >: A](index: Int, elem: B): View[B]^{this} = {
-    if(index < 0) throw new IndexOutOfBoundsException(index.toString)
-    val k = knownSize
-    if(k >= 0 && index >= k) throw new IndexOutOfBoundsException(index.toString)
-    iterableFactory.from(new View.Updated(this, index, elem))
-  }
 
   override def prepended[B >: A](elem: B): SeqView[B]^{this} = new SeqView.Prepended(elem, this)
   override def reverse: SeqView[A]^{this} = new SeqView.Reverse(this)

--- a/scala2-library-cc/src/scala/collection/SeqView.scala
+++ b/scala2-library-cc/src/scala/collection/SeqView.scala
@@ -34,12 +34,38 @@ trait SeqViewOps[+A, +CC[_], +C] extends Any with IterableOps[A, CC, C] {
   def reverse: C^{this}
   def sorted[B >: A](implicit ord: Ordering[B]): C^{this}
 
-  // Placeholder implementation for that method in SeqOps.
+  // Placeholder implementations for the corresponding methods in SeqOps.
   // This is needed due to the change in the class hierarchy in cc stdlib.
   // See #19660 and #19819.
+  // -------------------
   def updated[B >: A](index: Int, elem: B): CC[B]^{this} =
     assert(false, "This is a placeholder implementation in the capture checked Scala 2 library.")
     ???
+  def padTo[B >: A](len: Int, elem: B): CC[B]^{this} =
+    assert(false, "This is a placeholder implementation in the capture checked Scala 2 library.")
+    ???
+  def patch[B >: A](from: Int, other: IterableOnce[B]^, replaced: Int): CC[B]^{this, other} =
+    assert(false, "This is a placeholder implementation in the capture checked Scala 2 library.")
+    ???
+  def combinations(n: Int): Iterator[C^{this}]^{this} =
+    assert(false, "This is a placeholder implementation in the capture checked Scala 2 library.")
+    ???
+  def sortBy[B](f: A => B)(implicit ord: Ordering[B]): C^{this, f} =
+    assert(false, "This is a placeholder implementation in the capture checked Scala 2 library.")
+    ???
+  def sortWith(lt: (A, A) => Boolean): C^{this, lt} =
+    assert(false, "This is a placeholder implementation in the capture checked Scala 2 library.")
+    ???
+  def permutations: Iterator[C^{this}]^{this} =
+    assert(false, "This is a placeholder implementation in the capture checked Scala 2 library.")
+    ???
+  def distinct: C^{this} =
+    assert(false, "This is a placeholder implementation in the capture checked Scala 2 library.")
+    ???
+  def distinctBy[B](f: A -> B): C^{this} =
+    assert(false, "This is a placeholder implementation in the capture checked Scala 2 library.")
+    ???
+  // -------------------
 
   def reverseIterator: Iterator[A]^{this} = reversed.iterator
 }

--- a/scala2-library-cc/src/scala/collection/generic/IsSeq.scala
+++ b/scala2-library-cc/src/scala/collection/generic/IsSeq.scala
@@ -58,7 +58,8 @@ object IsSeq {
     * The helper method [[seqViewIsSeq_]] is added to make the binary compatible.
     */
   @annotation.targetName("seqViewIsSeq")
-  def seqViewIsSeq_[CC0[X] <: SeqView[X], A0]: IsSeq[CC0[A0]] { type A = A0; type C = View[A0] } = ???
+  @annotation.publicInBinary
+  private[IsSeq] def seqViewIsSeq_[CC0[X] <: SeqView[X], A0]: IsSeq[CC0[A0]] { type A = A0; type C = View[A0] } = ???
   implicit inline def seqViewIsSeq[CC0[X] <: SeqView[X], A0]: IsIterable[CC0[A0]] { type A = A0; type C = View[A0] } = seqViewIsSeq_[CC0, A0].asInstanceOf
 
   /** !!! Under cc, views are not Seqs and can't use SeqOps.
@@ -66,7 +67,8 @@ object IsSeq {
     * The helper method [[stringViewIsSeq__]] is added to make the binary compatible.
     */
   @annotation.targetName("stringViewIsSeq")
-  val stringViewIsSeq_ : IsSeq[StringView] { type A = Char; type C = View[Char] } = ???
+  @annotation.publicInBinary
+  private[IsSeq] val stringViewIsSeq_ : IsSeq[StringView] { type A = Char; type C = View[Char] } = ???
   inline implicit def stringViewIsSeq: IsIterable[StringView] { type A = Char; type C = View[Char] } = stringViewIsSeq_.asInstanceOf
 
   implicit val stringIsSeq: IsSeq[String] { type A = Char; type C = String } =

--- a/scala2-library-cc/src/scala/collection/generic/IsSeq.scala
+++ b/scala2-library-cc/src/scala/collection/generic/IsSeq.scala
@@ -54,24 +54,20 @@ object IsSeq {
     seqOpsIsSeqVal.asInstanceOf[IsSeq[CC0[A0]] { type A = A0; type C = CC0[A0] }]
 
   /** !!! Under cc, views are not Seqs and can't use SeqOps.
-   *  So this should be renamed to seqViewIsIterable
-   */
-  implicit def seqViewIsSeq[CC0[X] <: SeqView[X], A0]: IsIterable[CC0[A0]] { type A = A0; type C = View[A0] } =
-    new IsIterable[CC0[A0]] {
-      type A = A0
-      type C = View[A]
-      def apply(coll: CC0[A0]): IterableOps[A0, View, View[A0]] = coll
-    }
+    * Therefore, [[seqViewIsSeq]] now returns an [[IsIterable]].
+    * The helper method [[seqViewIsSeq_]] is added to make the binary compatible.
+    */
+  @annotation.targetName("seqViewIsSeq")
+  def seqViewIsSeq_[CC0[X] <: SeqView[X], A0]: IsSeq[CC0[A0]] { type A = A0; type C = View[A0] } = ???
+  implicit inline def seqViewIsSeq[CC0[X] <: SeqView[X], A0]: IsIterable[CC0[A0]] { type A = A0; type C = View[A0] } = seqViewIsSeq_[CC0, A0].asInstanceOf
 
   /** !!! Under cc, views are not Seqs and can't use SeqOps.
-   *  So this should be renamed to stringViewIsIterable
-   */
-  implicit val stringViewIsSeq: IsIterable[StringView] { type A = Char; type C = View[Char] } =
-    new IsIterable[StringView] {
-      type A = Char
-      type C = View[Char]
-      def apply(coll: StringView): IterableOps[Char, View, View[Char]] = coll
-    }
+    * Therefore, [[stringViewIsSeq]] now returns an [[IsIterable]].
+    * The helper method [[stringViewIsSeq__]] is added to make the binary compatible.
+    */
+  @annotation.targetName("stringViewIsSeq")
+  val stringViewIsSeq_ : IsSeq[StringView] { type A = Char; type C = View[Char] } = ???
+  inline implicit def stringViewIsSeq: IsIterable[StringView] { type A = Char; type C = View[Char] } = stringViewIsSeq_.asInstanceOf
 
   implicit val stringIsSeq: IsSeq[String] { type A = Char; type C = String } =
     new IsSeq[String] {

--- a/tests/run/enrich-gentraversable.scala
+++ b/tests/run/enrich-gentraversable.scala
@@ -3,6 +3,7 @@ import scala.language.postfixOps
 
 object Test extends App {
   import scala.collection.generic.IsIterable
+  import scala.collection.generic.IsSeq.seqViewIsSeq
   import scala.collection.{BuildFrom, Iterable, IterableOps, View}
   import scala.collection.immutable.TreeMap
 

